src/core/tiledscene/qgsgltfutils.cpp

```diff
@@ -0,0 +1,248 @@
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgltfutils.h"
+
+#include "qgscoordinatetransform.h"
+#include "qgsexception.h"
+#include "qgsmatrix4x4.h"
+
+#include <QImage>
+#include <QMatrix4x4>
+
+#define TINYGLTF_IMPLEMENTATION       // should be defined just in one CPP file
+
+// decompression of meshes with Draco is optional, but recommended
+// because some 3D Tiles datasets use it (KHR_draco_mesh_compression is an optional extension of GLTF)
+#ifdef HAVE_DRACO
+#define TINYGLTF_ENABLE_DRACO
+#endif
+
+#define TINYGLTF_NO_STB_IMAGE         // we use QImage-based reading of images
+#define TINYGLTF_NO_STB_IMAGE_WRITE   // we don't need writing of images
+//#define TINYGLTF_NO_FS
+
+//#include <fstream>
+#include "tiny_gltf.h"
+
+///@cond PRIVATE
+
+
+bool QgsGltfUtils::accessorToMapCoordinates( tinygltf::Model &model, int accessorIndex, const QgsMatrix4x4 &tileTransform, const QgsCoordinateTransform *ecefToTargetCrs, const QgsVector3D &tileTranslationEcef, QMatrix4x4 *nodeTransform, QVector<double> &vx, QVector<double> &vy, QVector<double> &vz )
+{
+  tinygltf::Accessor &accessor = model.accessors[accessorIndex];
+  tinygltf::BufferView &bv = model.bufferViews[accessor.bufferView];
+  tinygltf::Buffer &b = model.buffers[bv.buffer];
+
+  if ( accessor.componentType != TINYGLTF_PARAMETER_TYPE_FLOAT || accessor.type != TINYGLTF_TYPE_VEC3 )
+  {
+    // we may support more input types in the future if needed
+    return false;
+  }
+
+  unsigned char *ptr = b.data.data() + bv.byteOffset + accessor.byteOffset;
+
+  vx.resize( accessor.count );
+  vy.resize( accessor.count );
+  vz.resize( accessor.count );
+  for ( int i = 0; i < ( int )accessor.count; ++i )
+  {
+    float *fptr = reinterpret_cast<float *>( ptr );
+    QVector3D vOrig( fptr[0], fptr[1], fptr[2] );
+
+    if ( nodeTransform )
+      vOrig = nodeTransform->map( vOrig );
+
+    // go from y-up to z-up according to 3D Tiles spec
+    QVector3D vFlip( vOrig.x(), -vOrig.z(), vOrig.y() );
+
+    // apply also transform of the node
+    QgsVector3D v = tileTransform.map( QgsVector3D( vFlip ) + tileTranslationEcef );
+    vx[i] = v.x(); vy[i] = v.y(); vz[i] = v.z();
+
+    if ( bv.byteStride )
+      ptr += bv.byteStride;
+    else
+      ptr += 3 * sizeof( float );
+  }
+
+  if ( ecefToTargetCrs )
+  {
+    try
+    {
+      ecefToTargetCrs->transformCoords( accessor.count, vx.data(), vy.data(), vz.data() );
+    }
+    catch ( QgsCsException & )
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+std::unique_ptr<QMatrix4x4> QgsGltfUtils::parseNodeTransform( tinygltf::Node &node )
```

Comment:
should this be QgsMatrix4x4 ?

Reply:
Actually it was intentional to use QMatrix4x4. The rationale is that coordinates in GLTF should be generally small, and transform matrices as well (normally software would upload them as float matrices to GPU).

There's extra heuristic in QgsGltfUtils::extractTileTranslation() to extract translation that would not fit into floats without loss of precision, so that using the cheaper float math for matrix multiplications should be safe...

Reply:
Great! 👍